### PR TITLE
Implement aarch64 ABI for compiler-llvm.

### PR DESCRIPTION
**Pull request by nlewycky**
With this PR, `make llvm-test-native` passes on our arm64 machine.

### CODE REVIEW

File: lib/compiler-llvm/src/trampoline/wasm.rs [hunk 1]
```diff
@@ -286,225 +287,231 @@ impl FuncTrampoline {
...
+        let builder = context.create_builder();
+        builder.position_at_end(entry_block);
+
+        /*
+        // TODO: remove debugging
+        builder.build_call(
+            intrinsics.debug_trap,
+            &[],
+            "");
+        */
```
Comment (syrusakbary) [inline lib/compiler-llvm/src/trampoline/wasm.rs:309]: Remove this :)

Comment (nlewycky) [reply lib/compiler-llvm/src/trampoline/wasm.rs]: Done.

**Comment by syrusakbary** [comment]
Can we enable the tests for this?

**Comment by nlewycky** [comment]
> Can we enable the tests for this?

That's my next task. Because cranelift+jit is on, if I turn on llvm too it will try to test llvm+jit which does not work at all. If I turn off jit, and turn on llvm, we'll stop testing cranelift+jit.

**Comment by syrusakbary** [comment]
> That's my next task. Because cranelift+jit is on, if I turn on llvm too it will try to test llvm+jit which does not work at all. If I turn off jit, and turn on llvm, we'll stop testing cranelift+jit.

You mean to be done in the same PR on a separate one?

**Comment by nlewycky** [comment]
> You mean to be done in the same PR on a separate one?

In a separate one, if all goes according to #1 plan.

My backup plan is that if this PR hasn't landed by the time I have the Makefile change for the testing ready, then I'll make the other PR be "add individual testing of $(compiler_engine) configurations" and then this PR would add llvm-native to the testing.

**Comment by nlewycky** [comment]
> My backup plan is that if this PR hasn't landed by the time I have the Makefile change for the testing ready, then I'll make the other PR be "add individual testing of $(compiler_engine) configurations" and then this PR would add llvm-native to the testing.

That other PR is now out for review, #1720 . If that lands first then I'll update this PR to add testing of llvm-native to aarch64.

**Comment by nlewycky** [comment]
Testing is now enabled!

**Review by syrusakbary — APPROVED** [review]

File: lib/compiler-llvm/src/translator/abi.rs [hunk 2]
```diff
@@ -22,556 +21,1174 @@ use inkwell::{
...
+    fn pack_values_for_register_return<'ctx>(
+        &self,
+        intrinsics: &Intrinsics<'ctx>,
+        builder: &Builder<'ctx>,
+        values: &[BasicValueEnum<'ctx>],
+        func_type: &FunctionType<'ctx>,
+    ) -> Result<BasicValueEnum<'ctx>, CompileError>;
+}
+
+/// Implementation of the [`Abi`] trait for the AMD64 SystemV ABI.
```
Comment (syrusakbary) [inline lib/compiler-llvm/src/translator/abi.rs:85]: Can we divide this into different files?
* `lib/compiler-llvm/src/translator/abi/x86_64_systemv.rs`
* `lib/compiler-llvm/src/translator/abi/aarch64_systemv.rs`

Comment (nlewycky) [reply lib/compiler-llvm/src/translator/abi.rs]: Done.

File: lib/compiler-llvm/src/abi/mod.rs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+use aarch64_systemv::Aarch64SystemV;
+use x86_64_systemv::X86_64SystemV;
+
+pub fn get_abi(target_machine: &TargetMachine) -> Box<dyn Abi> {
+    if target_machine
+        .get_triple()
+        .as_str()
+        .to_string_lossy()
+        .starts_with("aarch64")
```
Comment (syrusakbary) [inline lib/compiler-llvm/src/abi/mod.rs:31]: Why not `get_cpu`?

https://thedan64.github.io/inkwell/inkwell/targets/struct.TargetMachine.html#method.get_cpu

Comment (nlewycky) [reply lib/compiler-llvm/src/abi/mod.rs:31]: I tried that! It returns the string `"generic"`.

**Comment by Hywan** [comment]
Impressive PR, well done!

**Pull request merged**